File: P5 2024-2/Paradigmas de Programacao/ListaExercicio01/Golang/Animal.go
package main

import (
	"fmt"
)

// Definindo a estrutura Triangulo
type Animal struct {
	especie string
	nome    string
}

func (a Animal) emitirSom() {
	fmt.Println(a.nome, "faz um barulho...")
}

type Cachorro struct {
	Animal
	especie string
}

func (c Cachorro) emitirSom() {
	fmt.Println("Au Au")
}

type Gato struct {
	Animal
	especie string
}

func (g Gato) emitirSom() {
	fmt.Println("Miau")
}

type Mamifero struct {
	Animal
}

func (m Mamifero) amamentar() {
	fmt.Println(m.nome, "está amamentando...")
}

type Ave struct {
	Animal
}

func (a Ave) voar() {
	fmt.Println(a.nome, "está voando...")
}

type Morcego struct {
	Ave
	Mamifero
}

type AnimalInterface interface {
	emitirSom()
}

type Zoologico struct {
	Animais []AnimalInterface
}

func (z *Zoologico) adicionarAnimal(a AnimalInterface) {
	z.Animais = append(z.Animais, a)
}

func (z *Zoologico) sonsAnimais() {
	fmt.Printf("O Zoológico tem %d animais\n", len(z.Animais))

	for i := range z.Animais {
		z.Animais[i].emitirSom()
	}
}

File: P5 2024-2/Paradigmas de Programacao/ListaExercicio01/Java/JavaProject/src/Animal.java
import java.util.ArrayList;
import java.util.List;

public class Animal {
    protected String especie;
    protected String nome;

    public Animal(String especie, String nome) {
        this.especie = especie;
        this.nome = nome;
    }

    public String getNome() {
        return nome;
    }

    public String emitirSom() {
        return "";
    }
}

class Cachorro extends Animal {
    public Cachorro(String nome) {
        super("Cachorro", nome);
    }
    @Override
    public String emitirSom() {
        System.out.println(this.nome + " diz: Au Au");
        return "Au Au";
    }
}
class Gato extends Animal {
    public Gato(String nome) {
        super("Gato", nome);
    }
    @Override
    public String emitirSom() {
        System.out.println(this.nome + " diz: Miau");
        return "Miau";
    }
}

class Zoologico {
    protected List<Animal> animais;

    public Zoologico() {
        this.animais = new ArrayList<Animal>();
    }
    public void addAnimal(Animal animal) {
        this.animais.add(animal);
    }

    public void getAnimais() {
        for (Animal animal : animais) {
            animal.emitirSom();
        }
    }
}